### PR TITLE
Get the node_map and full_nodes simultaneously in JIT for all backends

### PR DESCRIPTION
**Pull request by pavanky**
This needs to be benchmarked a little bit before getting merged.

### CODE REVIEW

**Comment by pavanky** [comment]
build arrayfire linux ci

**Comment by pavanky** [comment]
@umar456 this is ready, but I want to run a few benchmarks before this can be merged.

**Comment by pavanky** [comment]
@umar456 I think this is good to go. I removed the kernel launch  and measured how much time it takes for the black scholes example.

Here's the before:
```
ArrayFire v3.6.0 (OpenCL, 64-bit Linux, build 4a605715)
[0] AMD: Hawaii, 8163 MB
-1- NVIDIA: GeForce GTX 950, 1995 MB
-2- AMD: Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz, 15985 MB
** ArrayFire Black-Scholes Example **
**          by AccelerEyes         **

Warming up done
Input Data Size =   200000. Mean GPU Time: 0.174165 ms
Input Data Size =   400000. Mean GPU Time: 0.175247 ms
Input Data Size =   600000. Mean GPU Time: 0.174462 ms
Input Data Size =   800000. Mean GPU Time: 0.173966 ms
Input Data Size =  1000000. Mean GPU Time: 0.173661 ms
Input Data Size =  1200000. Mean GPU Time: 0.172599 ms
Input Data Size =  1400000. Mean GPU Time: 0.173715 ms
Input Data Size =  1600000. Mean GPU Time: 0.172764 ms
Input Data Size =  1800000. Mean GPU Time: 0.175066 ms
Input Data Size =  2000000. Mean GPU Time: 0.174304 ms
```

Here's the after:
```
[0] AMD: Hawaii, 8163 MB
-1- NVIDIA: GeForce GTX 950, 1995 MB
-2- AMD: Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz, 15985 MB
** ArrayFire Black-Scholes Example **
**          by AccelerEyes         **

Warming up done
Input Data Size =   200000. Mean GPU Time: 0.171428 ms
Input Data Size =   400000. Mean GPU Time: 0.166801 ms
Input Data Size =   600000. Mean GPU Time: 0.169735 ms
Input Data Size =   800000. Mean GPU Time: 0.179639 ms
Input Data Size =  1000000. Mean GPU Time: 0.169283 ms
Input Data Size =  1200000. Mean GPU Time: 0.169360 ms
Input Data Size =  1400000. Mean GPU Time: 0.166881 ms
Input Data Size =  1600000. Mean GPU Time: 0.168037 ms
Input Data Size =  1800000. Mean GPU Time: 0.169529 ms
Input Data Size =  2000000. Mean GPU Time: 0.169447 ms
```

So on average 0.169 ms vs 0.174 ms earlier.

**Comment by pavanky** [comment]
Using std::array made it even faster:

```
ArrayFire v3.5.1 (OpenCL, 64-bit Linux, build 5d9898b2)
[0] AMD: Hawaii, 8163 MB
-1- NVIDIA: GeForce GTX 950, 1995 MB
-2- AMD: Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz, 15985 MB
** ArrayFire Black-Scholes Example **
**          by AccelerEyes         **

Warming up done
Input Data Size =   200000. Mean GPU Time: 0.143233 ms
Input Data Size =   400000. Mean GPU Time: 0.143477 ms
Input Data Size =   600000. Mean GPU Time: 0.143162 ms
Input Data Size =   800000. Mean GPU Time: 0.142684 ms
Input Data Size =  1000000. Mean GPU Time: 0.141832 ms
Input Data Size =  1200000. Mean GPU Time: 0.146820 ms
Input Data Size =  1400000. Mean GPU Time: 0.153475 ms
Input Data Size =  1600000. Mean GPU Time: 0.142987 ms
Input Data Size =  1800000. Mean GPU Time: 0.143969 ms
Input Data Size =  2000000. Mean GPU Time: 0.153221 ms
```

File: src/backend/cuda/JIT/Node.hpp [hunk 1]
```diff
@@ -21,6 +22,7 @@ namespace cuda
 namespace JIT
 {
 
+    const int MAX_CHILDREN = 2;
```
Comment (umar456) [inline src/backend/cuda/JIT/Node.hpp]: static

**Comment by pavanky** [comment]
@umar456 squash this when merging.

**Pull request merged**